quickwit/quickwit-search/src/find_trace_ids_collector.rs

```diff
@@ -0,0 +1,508 @@
+                debug_assert!(found);
+                TraceIdFruit::new(&trace_id_bytes, trace_id.span_timestamp)
+            })
+            .collect()
+    }
+}
+
+struct SelectTraceIds {
+    num_traces: usize,
+    dedup_workbench: FnvHashMap<TermOrd, DateTime>,
+    select_workbench: Vec<TraceIdTermOrd>,
+    running_term_ord: Option<TermOrd>,
+    running_span_timestamp: DateTime,
+    span_timestamp_sentinel: DateTime,
+}
+
+impl SelectTraceIds {
+    fn new(num_traces: usize) -> Self {
+        Self {
+            num_traces,
+            dedup_workbench: FnvHashMap::with_capacity_and_hasher(
+                2 * num_traces,
+                Default::default(),
+            ),
+            select_workbench: Vec::with_capacity(2 * num_traces),
+            running_term_ord: None,
+            running_span_timestamp: DateTime::default(),
+            span_timestamp_sentinel: DateTime::from_timestamp_micros(i64::MIN),
+        }
+    }
+
+    fn collect(&mut self, term_ord: TermOrd, span_timestamp: DateTime) {
+        if self.running_term_ord.is_none() {
+            self.running_term_ord = Some(term_ord);
+            self.running_span_timestamp = span_timestamp;
+            return;
+        }
+        if self.span_timestamp_sentinel >= span_timestamp {
+            return;
+        }
+        let running_term_ord = self
+            .running_term_ord
+            .expect("The running trace ID should be set.");
+
+        if running_term_ord == term_ord {
+            self.running_span_timestamp = self.running_span_timestamp.max(span_timestamp);
+        } else {
+            self.dedup(running_term_ord, self.running_span_timestamp);
+            self.truncate();
+            self.running_term_ord = Some(term_ord);
+            self.running_span_timestamp = span_timestamp;
+        }
+    }
+
+    #[cfg(any(test, feature = "testsuite"))]
+    fn collect_for_test(&mut self, term_ord: TermOrd, span_timestamp_micros: i64) {
+        let span_timestamp = DateTime::from_timestamp_micros(span_timestamp_micros);
+        self.collect(term_ord, span_timestamp)
+    }
+
+    fn dedup(&mut self, term_ord: TermOrd, span_timestamp: DateTime) {
+        self.dedup_workbench
+            .entry(term_ord)
+            .and_modify(|entry| {
+                if *entry < span_timestamp {
+                    *entry = span_timestamp
+                }
+            })
+            .or_insert(span_timestamp);
+    }
+
+    fn select(&mut self) {
+        if self.dedup_workbench.is_empty() {
+            return;
+        }
+        self.select_workbench.clear();
+
+        for (term_ord, span_timestamp) in self.dedup_workbench.drain() {
+            let trace_id = TraceIdTermOrd::new(term_ord, span_timestamp);
+            self.select_workbench.push(trace_id);
+        }
+        let select_len = self.num_traces.min(self.select_workbench.len());
+        let select_index = select_len - 1;
+        self.select_workbench.select_nth_unstable(select_index);
```

Comment:
TIL that it uses a version of quickselect/quicksort that does not exhibit the worst case nightmares of quicksort.